src/utils/auth_token/org_auth_token.rs

```diff
@@ -0,0 +1,89 @@
+use super::{AuthTokenParseError, Result};
+use serde::Deserialize;
+
+const ORG_AUTH_TOKEN_PREFIX: &'static str = "sntrys_";
+const ORG_TOKEN_SECRET_BYTES: usize = 32;
+
+/// Represents a valid org auth token.
+#[derive(Debug, Clone)]
+pub struct OrgAuthToken {
+    auth_string: String,
+    pub payload: AuthTokenPayload,
+}
+
+/// Represents the payload data of an org auth token.
+#[derive(Clone, Debug, Deserialize)]
+#[allow(dead_code)] // Otherwise, we get a warning about unused fields
+pub struct AuthTokenPayload {
+    iat: f64,
+    pub url: Option<String>, // URL may be missing from some old auth tokens, see getsentry/sentry#57123
+    region_url: String,
+    pub org: String,
+}
+
+impl OrgAuthToken {
+    /// Parses the payload of an org auth token from the encoded payload string extracted from the token
+    /// string. Returns an error if the payload cannot be decoded or if the deceded payload is not valid JSON.
+    fn parse_payload(payload_encoded: &str) -> Result<AuthTokenPayload> {
+        let payload_bytes = data_encoding::BASE64
+            .decode(payload_encoded.as_bytes())
+            .map_err(|_| AuthTokenParseError)?;
+
+        let payload = String::from_utf8(payload_bytes).map_err(|_| AuthTokenParseError)?;
+
+        serde_json::from_str(&payload).map_err(|_| AuthTokenParseError)
+    }
+
+    /// Validates the secret segment of an org auth token. Returns an error if the secret is not valid.
+    fn validate_secret(secret: &str) -> Result<()> {
+        let num_bytes = data_encoding::BASE64_NOPAD
+            .decode(secret.as_bytes())
+            .map(|bytes| bytes.len());
+
+        match num_bytes {
+            Ok(ORG_TOKEN_SECRET_BYTES) => Ok(()),
+            _ => Err(AuthTokenParseError),
+        }
+    }
+
+    /// Constructs a new OrgAuthToken from a string. Returns an error if the string is not a valid org auth token.
+    fn construct_from_string(auth_string: String) -> Result<OrgAuthToken> {
+        if !auth_string.starts_with(ORG_AUTH_TOKEN_PREFIX) {
+            return Err(AuthTokenParseError);
+        }
+
+        let mut segment_split = auth_string.split('_');
+        segment_split.next(); // Skip the prefix
```

Comment:
you should probably still validate it though

Reply:
It is already validated on lines 51-53 – I will update the comment to indicate that we are skipping since the prefix is already validated to avoid future confusion